### ISSUE TITLE
[RML]_LR.R

### ISSUE THREAD

**Issue by takytaky** [takytaky/rusita#5]
# 작업디렉토리 지정 

getwd()
setwd("~/[005]_LR")
getwd()

# 예측모델(Regression) Review ####

R_Machine <- function(x, w, b){
  
  y_hat <- x * w + b
  
  return(y_hat)
}

# R_Machine(2, 3, 4)
# R_Machine(c(3,6,9), 2, 3)

# Cost Function

Cost <- function(x, y, w, b){
  
  y_hat <- R_Machine(x, w, b)
  
  loss <- (y - y_hat)^2
  
  cost <- mean(loss)
  
  return(cost)
}

# 경사하강법 
# w변화(0.001)와 b변화(0.001)에 따른 
# Cost 변화량 계산

Trainer <- function(x, y, w, b, step){
  
  dw <- (Cost(x, y, w + 0.0001, b) - Cost(x, y, w, b)) / 0.0001
  db <- (Cost(x, y, w, b + 0.0001) - Cost(x, y, w, b)) / 0.0001
  
  w <- w - step * dw
  b <- b - step * db
  
  return(c(w, b))
}

# Testing

set.seed(2045)

# x, y 난수값 생성

x <- rnorm(20)
y <- 2 * x + 1 + rnorm(20) * 0.01

# 초기값 지정

w <- 1
b <- 0

# 1만번 학습

for(i in 1:10000){
  temp <- Trainer(x, y, w, b, 0.01)
  
  w <- temp[1]
  b <- temp[2]
}

# 결과 확인

w
b

# 함수로 확인

lm(y ~ x)

# 분류모델(Classification) ####

# sigmoid( ) 함수 지정

sigmoid <- function(x){
  
  y <- exp(x) / (1 + exp(x))
  
  return(y)
}

# sigmoid( ) 테스트

sigmoid(0)
sigmoid(100)
sigmoid(-1000)

# sigmoid( ) 함수를 필터로 적용
# C_Machine( ) 분류머신 생성

C_Machine <- function(x, w, b){
  
  y_hat <- sigmoid(w * x + b)
  
  return(y_hat)
}

# Cost Function -> Cross-Entropy

# Information Theory(정보 이론)
# Information Gain(정보 이득량)
# I(x) = -log(P(x))
#  -> 자주 발생하지 않는 사건은
#     자주 발생하는 사건보다 정보량이 많음

# Ex_1) 정보의 희귀성에 비례
#       A보다 B, C의 정보량이 더 높음

A <- 0.6 ; B <- 0.3 ; C <- 0.1
(I_A <- -log(A))
(I_B <- -log(B))
(I_C <- -log(C))

# Ex_2) Degree of Surprise(Information)
#       AlphaGo Vs. Apes

AlphaGo <- 0.99 ; Apes <- 0.01
(I_AlphaGo <- -log(AlphaGo))
(I_Apes <- -log(Apes))

# Entropy(불확실성 척도) : Entropy = E(-log(P(x)))
# 정의 : 정보량들의 평균
# Entropy가 낮으면 불확실성이 낮음
# 불확실성이 낮으면 분류정확도가 높아짐

# (승리)확률변수 P1과 P2를 사용하여 엔트로피 계산

# 001) Entropy : 0.056(실력 차이가 많은 경우)

P1 <- 0.99 ; P2 <- 0.01
(E1 <- -P1 * log(P1) - P2 * log(P2))

# 002) Entropy : 0.693(실력 차이가 적은 경우)

P1 <- 0.5 ; P2 <- 0.5
(E2 <- -P1 * log(P1) - P2 * log(P2))

# E1(0.056) vs. E2(0.693)
# Entropy : 1) 확률변수의 평균정보량
#           2) Surprise의 평균정도
#           3) 불확실성의 정도

E1 ; E2

# KL(Kullback-Leibler)-Divergence/KLD
# Relative Entropy
# Q(x):예측확률, P(x):실제확률
# KLD = E(-log(Q(x))) - E(-log(P(x)))
# KLD값은 항상 0보다 크고
# 예측확률이 실제확률에 가까와 질수록 0에 가까워짐

# Cross-Entropy
# y    :실제값 -> 지도학습(교사데이터)
# y_hat:예측값 -> Model 적용값

# 교사데이터(y)가 Cross-Entropy의 가중치로 적용
# CE(-log(y_hat)) = y * -log(y_hat)
#                 = -y * log(y_hat)

# Loss = -y * log(y_hat) - (1 - y) * log(1 - y_hat)
# Cost = sum(Loss)

Cost <- function(x, y, w, b){
  
  y_hat <- C_Machine(x, w, b)
  
  loss <- -y * log(y_hat) - (1 - y) * log(1 - y_hat)
  
  cost <- mean(loss)
  
  return(cost)
}

# 경사하강법 - 분류모델
# w변화(0.001)와 b변화(0.001)에 따른 
# Cost 변화량 계산

Trainer <- function(x, y, w, b, step){
  
  dw <- (Cost(x, y, w + 0.0001, b) - Cost(x, y, w, b)) / 0.0001
  db <- (Cost(x, y, w, b + 0.0001) - Cost(x, y, w, b)) / 0.0001
  
  w <- w - step * dw
  b <- b - step * db
  
  return(c(w, b))
}

# Default Data Classification

# Default 데이터 읽어오기

Default <- read.csv("data/Default.csv")

str(Default)

# Default(카드연체), 학생여부, 잔고, 수입

head(Default)

table(Default$default)

plot(Default$balance ~ Default$default)

# One-hot Encoding

y <- ifelse(Default$default == "Yes", 1, 0)

table(y)

# balance 범위 확인

BLC <- Default$balance

range(BLC)

plot(BLC, cex = 0.5)

boxplot(BLC)

# Normalization

range(BLC / sd(BLC))

x <- Default$balance / sd(Default$balance)

range(x)

# 결과 비교

par(mfrow = c(2,2))

# Before Normalization

plot(BLC, cex = 0.5)

boxplot(BLC)

# After Normalization

plot(x, cex = 0.5)

boxplot(x)

par(mfrow = c(1,1))

# 경사하강법
# Trainer() Model 적용

w <- 0
b <- 0

# 반복횟수(5000)와 학습률(0.5) 값을 변경하면서 확인!!!
# 2분정도 소요

for(i in 1:5000){
  temp <- Trainer(x, y, w, b, 0.5)
  
  w <- temp[1]
  b <- temp[2]
}

w
b

# 업데이트된 파라미터 적용
# 0 ~ 1 사이 확률값

y_prob <- C_Machine(x, w, b)

head(y_prob)

# 0.5(변경가능) 기준 분류
# 분류예측

y_label <- ifelse(y_prob > 0.5, 1, 0)

table(y_label)

# 실제값과 예측값을 비교
# 이진 혼돈 행렬(Binary Confusion Matrix)

table(Default$default, y_label)

# Model Validation(상환기준)

(accuracy <- (9633 + 96)/(9633 + 34 + 237 + 96))

(precision <- 9633/(9633 + 237))

(recall <- 9633/(9633 + 34))

# F-Score

F_Score <- 2 * ((precision * recall) / (precision + recall))

F_Score

# Model Validation(연체기준)

(accuracy <- (9633 + 96)/(9633 + 34 + 237 + 96))

(precision <- 96/(34 + 96))

(recall <- 96/(237 + 96))

# F-Score

F_Score <- 2 * ((precision * recall) / (precision + recall))

F_Score

# glm() 함수 적용 ####

# Modeling

Model_lr <-  glm(formula = y ~ x, family = "binomial")

Model_lr

# Model_lr Model 적용
# 0 ~ 1 사이의 확률값

y_prob2 <- predict(Model_lr, type = "response")

head(y_prob2)

# 0.5기준으로 분류

y_label2 <- ifelse(y_prob2 > 0.5, 1, 0)

# 이진 혼돈 행렬(Binary Confusion Matrix)

table(y, y_label2)

# Model Validation

(precision2 <- 100/(42 + 100))

(recall2 <- 100/(233 + 100))

# F-Score2

F_Score2 <- 2 * ((precision2 * recall2) / (precision2 + recall2))

F_Score2

# ROC_AUC
# install.packages("ROCR")

library(ROCR)

# prediction(예측값, 실제값)

y_prob2 <- prediction(y_label2, y)

# performance(prediction() 생성객체, 
#             measure = y축, 
#             x.measure = x축)
# x축(fpr) : False Positive Rate
# y축(tpr) : True Positive Rate

AUC_lr <- performance(y_prob2, 
                      measure = "tpr", 
                      x.measure = "fpr")

# performance()로 생성한 객체를 시각화

plot(AUC_lr)

# ranger 패키지 ####
# install.packages("ranger")

library(ranger)

# Modeling

Model_ranger <- ranger(default ~ balance, 
                       data = Default, 
                       probability = TRUE)

# 0 ~ 1 사이 확률값

y_prob3 <- Model_ranger$predictions[, 2]

head(y_prob3, 100)

# 0.5 기분 분류

y_label3 <- ifelse(y_prob3 > 0.5, 1, 0)

# 교차빈도표

table(y, y_label3)

# Model Validation

(recall3 <- 108/(225 + 108))

(precision3 <- 108/(161 + 108))

# F-Score -> P-R Curve & PRAUC

F_Score3 <- 2 * ((precision3 * recall3) / (precision3 + recall3))

F_Score3

# Smarket Data ####
# 야후파이낸스 S&P 500 Index
# Year      : 거래 년도
# Lag1~5    : 1~5일전 수익률
# Volume    : 주식거래량
# Today     : 오늗의 수익률
# Direction : 상승(Up), 하락(Down)

Smarket <- read.csv("data/Smarket.csv")

str(Smarket)

head(Smarket)

# Modeling

Model_SM <- glm(Direction ~ Lag1 + Lag2 + Lag3 + Lag4 + Lag5 + Volume,
            data = Smarket, 
            family = binomial)

# 변수 유효성

summary(Model_SM)

# 0 ~ 1 사이의 확률값으로 적용

Pred_PB <- predict(Model_SM, type = "respons")

# 결과 5개 확인

head(Pred_PB)

# 0.5를 기준으로 "Up" 또는 "Down" 분류

Pred_LR <- ifelse(Pred_PB > 0.5, "UP", "Down")

# 분류결과 일부 확인

head(Pred_LR)

# 교차빈도표 생성

attach(Smarket)

table(Pred_LR, Direction)

# Model Validation

(recall_S <- 507/(457 + 507))

(precision_S <- 507/(141 + 507))

# F-Score -> P-R Curve & PRAUC

F_Score_S <- 2 * ((precision_S * recall_S) / (precision_S + recall_S))

F_Score_S

# Multinomial Logistic Regression ####
# install.packages("nnet")

# Neural Network Package

library(nnet)

# iris Data

iris <- read.csv("data/IRIS.csv")

# Modeling

Model_MLR <- multinom(Species ~ ., data = iris)

# 확률값 확인

head(fitted(Model_MLR))

tail(head(fitted(Model_MLR), 80))

tail(fitted(Model_MLR))

# 분류에 포함될 확률값(type = "probs")

Pred_MLR_P <- predict(Model_MLR, newdata = iris, type = "probs")

head(Pred_MLR_P)

# 분류값(type = "class")

Pred_MLR_C <- predict(Model_MLR, newdata = iris, type = "class")

head(Pred_MLR_C)

# 교차빈도표

xtabs(~ Pred_MLR_C + iris$Species)
